src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -498,6 +498,29 @@ public List<TSource> ToList()
                 return list;
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                int count = GetCount(onlyIfCheap: true);
+                var hashSet = count != -1 ? new HashSet<TSource>(count, comparer) : new HashSet<TSource>(comparer);
+
+                for (int i = 0; ; i++)
```

Comment:
Maybe start with `new HashSet<TSource>(GetEnumerable(0))` and then add from `1` on?